doc/rtd/reference/datasources/nocloud.rst

```diff
@@ -0,0 +1,190 @@
+.. _datasource_nocloud:
+
+NoCloud
+*******
+
+The data source ``NoCloud`` allows the user to provide user data and metadata
+to the instance without running a network service (or even without having a
+network at all).
+
+You can provide metadata and user data to a local VM boot via files on a
+`vfat`_ or `iso9660`_ filesystem. The filesystem volume label must be
+``cidata`` or ``CIDATA``.
+
+Alternatively, you can provide metadata via the kernel command line or SMBIOS
+"serial number" option. The data must be passed in the form of a string: ::
+
+  ds=nocloud[;key=val;key=val]
+
+or, ::
+
+  ds=nocloud-net[;key=val;key=val]
+
+Permitted keys
+==============
+
+The permitted keys are:
+
+* ``h`` or ``local-hostname``
+* ``i`` or ``instance-id``
+* ``s`` or ``seedfrom``
+
+With ``ds=nocloud``, the ``seedfrom`` value must start with ``/`` or
+``file://``.  With ``ds=nocloud-net``, the ``seedfrom`` value must start
+with ``http://`` or ``https://``.
+
+For example, you can pass this option to QEMU: ::
+
+  -smbios type=1,serial=ds=nocloud-net;s=http://10.10.0.1:8000/
+
+This will cause NoCloud to fetch the full metadata from
+http://10.10.0.1:8000/meta-data after the network initialisation is complete.
+
+File formats
+============
+
+These user data and metadata files are expected to be in the following
+format: ::
+
+  /user-data
+  /meta-data
+
+Both files must be present for it to be considered a valid seed ISO.
+
+Basically, ``user-data`` is simply user data and ``meta-data`` is a
+YAML-formatted file representing what you'd find in the EC2 metadata service.
+
+You may also optionally provide a vendor data file in the following format: ::
+
+  /vendor-data
+
+Creating a disk
+===============
+
+Given a disk Ubuntu cloud image in :file:`disk.img`, you can create a
+sufficient disk by following the following example.
+
+1. Create the :file:`user-data` and :file:`meta-data` files that will be used
+   to modify the image on first boot.
+
+.. code-block:: sh
+
+   $ echo -e "instance-id: iid-local01\nlocal-hostname: cloudimg" > meta-data
+   $ echo -e "#cloud-config\npassword: passw0rd\nchpasswd: { expire: False }\nssh_pwauth: True\n" > user-data
+
+2. At this stage you have three options:
+
+   a. Create a disk to attach with some user data and metadata:
```

Comment:
Unfortunately I'm not sure how to change it to not be like that. If I take the code blocks outside of the list structure, the list elements start to have the same indentation level as the code, and so become part of the code blocks. It's very annoying, but I'm not sure if there's a way around it.

Reply:
One idea is use a drop down to hide the other options, which would keep the preferred options all on the same indentation level (and reduce the amount of required visual parsing).

Reply:
Since we haven't used dropdowns elsewhere, I think I'd like to avoid that option. It might be better to put a pin in this one and revise the steps from a content POV so that the indentation becomes unnecessary. I think what might be better is if we get "opinionated" on this sequence and recommend a particular path, then provide the other options as alternatives below. If we can tolerate the visual oddity for now, that is!

Reply:
Works for me.